### PR TITLE
video_core: Refactor GPU interface

### PR DESCRIPTION
**Pull request by GPUCode**
The current video_core organization is a mess of global state that desperately needed some cleanup. This PR improves the structure of the project, removing the vast majority of global state and improving accuracy at the same time.

### Why this was needed
On master GPU registers were split between the external ones present in core/hw/gpu.h and the internal PICA registers in video_core/regs.h. This was problematic for a couple of reasons. Command buffer registers had a separate copy for both of these sets of registers but as noted on the relevant 3dbrew [page](https://www.3dbrew.org/wiki/GPU/External_Registers#Map) they are the same registers as the [internal ones](https://www.3dbrew.org/wiki/GPU/Internal_Registers#GPUREG_CMDBUF_SIZE0). Because of that, there was also excessive padding to get the correct offsets. Another problem was that renderers need access to external GPU framebuffer registers to configure the display output so they were global state which is always not ideal.

The solution to this, is to unify the PICA register map in a single class that contains all the registers in their correct offsets. During this transition I've also fixed a couple of incorrectly placed structures. However the problem now is how GSP will access those registers to trigger the necessary GPU commands. We ideally don't want to expose video_core implementation details to core services, that would be considered bad design, which leads me to the next point.

### The new interface

Access to video_core is now gated behind the GPU class which uses pimpl to hide its implementation details. It manages the renderer plugin, signalling GSP interrupts and handling all reads and writes to the PICA register map. Command list handling is now in a separate class called PicaCore which is a combination of CommandProcessor and State from master.

The shader.h header which was quite large before has been split up into its logical components, namely the shader_unit.h which exposes the shader unit classes, the shader_setup.h header which exposes common state between the shader units and output_vertex.h. With this distiction users can include only the declarations they wish instead of bringing a large header each time.

For better organization all classes with the Pica namespace have been moved into the pica subfolder, so the root directory is free to house common renderer interfaces (for example future common shader cache implementation will go there). In addition (almost) all video core global state has been eliminated including g_state, g_renderer and g_memory. It was attempted to also get rid of g_debug_context but the Qt debug widgets expect this to be available before system has intiailized so I left it as is to prevent blowing up the PR as big as it is already.

### Accuracy improvements.

The extra parameters field in CommandHeader has been changed to 8bit according to [3dbrew](https://www.3dbrew.org/wiki/GPU/Internal_Registers#Command_Header) which improved a bunch of games that had invalid headers in command lists (see below).

Triggering command buffer execution now correctly uses GPUREG_CMDBUF_ADDR0/GPUREG_CMDBUF_SIZE0 instead of the unrelated registers. The unification of the register map also ensures that writes to internal registers with WriteHWRegs works as expected, as its technically not disallowed for games to do that.

Revival of https://github.com/citra-emu/citra/pull/5342 which I have confirmed to be correct based on GSP module disassembly. I don't know if this fixes any games but we should be accurate here. I've also done a bunch of small cleanups in the GSP module and de-globalized it, since interrupt handling is now performed by the GPU class with a callback.

The above improvements have fixed Mini Mario & Friends, Mario vs. Donkey Kong: Tipping Stars and Mario Golf - World Tour which no longer softlock and have missing graphics.

Citra (master) | Citra (PR)
-|-
bad | good
bad_tipp | good_tipp

Fixes
https://github.com/citra-emu/citra/issues/3183
https://github.com/citra-emu/citra/issues/4712
https://github.com/citra-emu/citra/issues/6985

Supersedes
https://github.com/citra-emu/citra/pull/5342

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7272)

### CODE REVIEW

File: src/video_core/pica/output_vertex.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                                                                    "offset.")
+
+ASSERT_POS(pos, RasterizerRegs::VSOutputAttributes::POSITION_X);
+ASSERT_POS(quat, RasterizerRegs::VSOutputAttributes::QUATERNION_X);
+ASSERT_POS(color, RasterizerRegs::VSOutputAttributes::COLOR_R);
+ASSERT_POS(tc0, RasterizerRegs::VSOutputAttributes::TEXCOORD0_U);
+ASSERT_POS(tc1, RasterizerRegs::VSOutputAttributes::TEXCOORD1_U);
+ASSERT_POS(tc0_w, RasterizerRegs::VSOutputAttributes::TEXCOORD0_W);
+ASSERT_POS(view, RasterizerRegs::VSOutputAttributes::VIEW_X);
+ASSERT_POS(tc2, RasterizerRegs::VSOutputAttributes::TEXCOORD2_U);
```
Comment (Steveice10) [inline src/video_core/pica/output_vertex.cpp:46]: Is there a reason these specific static asserts happen in the cpp file when we usually put them with the structure in the header? There could be a situation where a specific compilation unit has settings that mess with the offsets.

File: src/video_core/pica/regs_external.cpp
Comment (Steveice10) [inline src/video_core/pica/regs_external.cpp:1]: Empty file?

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 2]
```diff
@@ -30,7 +30,7 @@ enum class PrecompiledEntryKind : u32 {
     Dump,
 };
 
-constexpr u32 NativeVersion = 1;
+constexpr u32 NativeVersion = 2;
```
Comment (Steveice10) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:33]: Are there any specific changes affecting the transferable cache that warranted this version bump? As this means everyone's shader caches will be reset to absolutely empty, including transferable raw shader binaries.

**Review by Steveice10 — COMMENTED** [review]
Looks great, thanks for doing this. Just a few things from me.

**Comment by Steveice10** [comment]
Got this attempting to open Mario Kart 7 with the PR artifact, does not happen on master:
```
libc++abi: terminating due to uncaught exception of type oaknut::OaknutException: out of range
zsh: abort      ./citra-qt.app/Contents/MacOS/citra-qt
```

Trace:
```
Thread 43 Crashed:: EmuThread
0   libsystem_kernel.dylib        	       0x18937a0dc __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x1893b1cc0 pthread_kill + 288
2   libsystem_c.dylib             	       0x1892bda40 abort + 180
3   libc++abi.dylib               	       0x189369070 abort_message + 132
4   libc++abi.dylib               	       0x189359110 demangling_terminate_handler() + 320
5   libobjc.A.dylib               	       0x188fff99c _objc_terminate() + 160
6   libc++abi.dylib               	       0x189368434 std::__terminate(void (*)()) + 16
7   libc++abi.dylib               	       0x18936b520 __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 88
8   libc++abi.dylib               	       0x18936b464 __cxa_throw + 308
9   citra-qt                      	       0x101535118 Pica::Shader::JitShader::Compile(std::__1::array<unsigned int, 4096ul> const*, std::__1::array<unsigned int, 4096ul> const*) + 192
10  citra-qt                      	       0x101534540 Pica::Shader::JitEngine::SetupBatch(Pica::ShaderSetup&, unsigned int) + 492
11  citra-qt                      	       0x101431ef4 0x100714000 + 13754100
12  citra-qt                      	       0x101431be4 Pica::PicaCore::WriteInternalReg(unsigned int, unsigned int, unsigned int) + 1092
13  citra-qt                      	       0x10143177c Pica::PicaCore::ProcessCmdList(unsigned int, unsigned int) + 324
14  citra-qt                      	       0x10141df80 0x100714000 + 13672320
15  citra-qt                      	       0x10141d870 VideoCore::GPU::Execute(Service::GSP::Command const&) + 1448
16  citra-qt                      	       0x101232c30 0x100714000 + 11660336
17  citra-qt                      	       0x1010ba9f4 Kernel::ServerSession::HandleSyncRequest(std::__1::shared_ptr<Kernel::Thread>) + 896
18  citra-qt                      	       0x101061b88 Kernel::ClientSession::SendSyncRequest(std::__1::shared_ptr<Kernel::Thread>) + 156
19  citra-qt                      	       0x1010caeac 0x100714000 + 10186412
20  citra-qt                      	       0x1010d9630 0x100714000 + 10245680
21  citra-qt                      	       0x1010d8498 Kernel::SVC::CallSVC(unsigned int) + 304
22  ???                           	       0x3117858e0 ???
23  citra-qt                      	       0x1013bda24 0x100714000 + 13277732
24  citra-qt                      	       0x100fb6740 Core::System::RunLoop(bool) + 1780
25  citra-qt                      	       0x100e40154 EmuThread::run() + 684
26  QtCore                        	       0x10d26ffb4 QThreadPrivate::start(void*)::$_0::operator()() const + 264 (qthread_unix.cpp:324) [inlined]
27  QtCore                        	       0x10d26ffb4 void (anonymous namespace)::terminate_on_exception<QThreadPrivate::start(void*)::$_0>(QThreadPrivate::start(void*)::$_0&&) + 264 (qthread_unix.cpp:260) [inlined]
28  QtCore                        	       0x10d26ffb4 QThreadPrivate::start(void*) + 336 (qthread_unix.cpp:283)
29  libsystem_pthread.dylib       	       0x1893b2034 _pthread_start + 136
30  libsystem_pthread.dylib       	       0x1893ace3c thread_start + 8
```

File: src/video_core/geometry_pipeline.cpp [hunk 3]
```diff
@@ -49,60 +51,63 @@ class GeometryPipelineBackend {
...
+    const RegsInternal& regs;
+    GeometryShaderUnit& unit;
+    AttributeBuffer attribute_buffer;
     Common::Vec4<f24>* buffer_cur;
     Common::Vec4<f24>* buffer_end;
-    unsigned int vs_output_num;
+    u32 vs_output_num;
 
-    GeometryPipeline_Point() : regs(g_state.regs), unit(g_state.gs_unit) {}
+    // TODO: REMOVE THIS
```
[the remark this answers is not in the document]
Comment (GPUCode) [reply src/video_core/geometry_pipeline.cpp:95]: This is needed sadly by boost serialization, so we can't remove this until we switch to a custom serialization solution

File: src/video_core/pica/output_vertex.cpp (hunk 1 above)
Comment (GPUCode) [reply src/video_core/pica/output_vertex.cpp:46]: Put it in cpp file to prevent needing to include regs_rasterizer in the header just for these asserts. There aren't any macros anyway that can change the layout of the structure so it should be fine imo

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp (hunk 2 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:33]: Bumped it just as a preventative measure since the register map was updated. Though since the internal registers were not changed it should be okay to not bump this actually

File: src/citra_qt/debugger/graphics/graphics.cpp [hunk 4]
```diff
@@ -75,11 +75,15 @@ GPUCommandStreamWidget::GPUCommandStreamWidget(QWidget* parent)
 }
 
 void GPUCommandStreamWidget::showEvent(QShowEvent* event) {
-    g_debugger.RegisterObserver(&model);
+    if (system.IsPoweredOn()) {
+        system.GPU().Debugger().RegisterObserver(&model);
```
[the remark this answers is not in the document]
Comment (GPUCode) [reply src/citra_qt/debugger/graphics/graphics.cpp:79]: Yeah thinking back this can break in many ways. Need a better way to make sure this abides by the lifetime of the gpu 🤔

**Comment by Dragios** [comment]
Should close #6985 too

File: src/core/memory.cpp [hunk 5]
```diff
@@ -346,13 +345,52 @@ std::shared_ptr<PageTable> MemorySystem::GetCurrentPageTable() const {
     return impl->current_page_table;
 }
 
+void RasterizerFlushVirtualRegion(VAddr start, u32 size, FlushMode mode) {
+    VAddr end = start + size;
```
Comment (FearlessTobi) [inline src/core/memory.cpp:349]:
```suggestion
   const VAddr end = start + size;
```

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 6]
```diff
@@ -302,23 +286,21 @@ void RendererOpenGL::LoadFBToScreenInfo(const GPU::Regs::FramebufferConfig& fram
...
 
     // Update existing texture
-    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, 1, 1, 0, GL_RGB, GL_UNSIGNED_BYTE, framebuffer_data);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, 1, 1, 0, GL_RGB, GL_UNSIGNED_BYTE, color.AsArray());
 
     state.texture_units[0].texture_2d = 0;
     state.Apply();
+
+    // Resize the texture in case the framebuffer size has changed
+    texture.width = 1;
```
Comment (FearlessTobi) [inline src/video_core/renderer_opengl/renderer_opengl.cpp:302]: Why do these have the size one?

File: src/video_core/renderer_vulkan/pica_to_vk.h [hunk 7]
```diff
@@ -56,14 +54,6 @@ inline vk::SamplerAddressMode WrapMode(Pica::TexturingRegs::TextureConfig::WrapM
 
     const auto index = static_cast<std::size_t>(mode);
     ASSERT_MSG(index < wrap_mode_table.size(), "Unknown texture wrap mode {}", index);
-
-    if (index > 3) {
```
Comment (FearlessTobi) [inline src/video_core/renderer_vulkan/pica_to_vk.h:60]: why is this removed?

**Review by FearlessTobi — COMMENTED** [review]
Looks very good, thanks for fixing these issues

**Comment by Wunkolo** [comment]
> Got this attempting to open Mario Kart 7 with the PR artifact, does not happen on master. Happens during the spinning wheel on the loading screen. Hardware shader is enabled. Does not happen if shader JIT is disabled.

This is due to the structure-offset being too far for a particular ARM64 instruction to address into.
I ran into this with usage of `offsetof` for serializing certain shader-unit states. For example:
https://github.com/citra-emu/citra/blob/f177433d41e3434b9530129283544d5b66cf85a4/src/video_core/shader/shader_jit_a64_compiler.cpp#L631-L637

The remedy would be to just move these variables farther up in the `UnitState` structure so that the offset would be lower.

**Comment by Steveice10** [comment]
Yeah we figured it out and it’s fixed now, thanks.

File: src/video_core/renderer_vulkan/pica_to_vk.h (hunk 7 above)
Comment (GPUCode) [reply src/video_core/renderer_vulkan/pica_to_vk.h:60]: To get rid of unnecessary global state. And I dont think we have access to telemetry anymore

File: src/video_core/renderer_opengl/renderer_opengl.cpp (hunk 6 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/renderer_opengl.cpp:302]: When the screen is filled, opengl will recreate the texture as 1x1 and stretch it across the screen.

**Comment by GPUCode** [comment]
> This seems to break everything for me on Android

Could you upload a log file with the crash, might show why it broke

**Comment by Dragios** [comment]
I'm still seeing command list error log and the game slows down significantly during log spam.
Performance is way better than master though and missing graphic is definitely fixed.
Is this intended?

The game actually did not soft lock on master but slow down to a crawl due to log spam. Once log filter downgraded to `:Critical`, the game run smooth.

Game tested was _Mini Mario and Friends_.

**Comment by GPUCode** [comment]
I tested mini mario and friends and it ran smooth on slopes. Can you upload a log file?

**Comment by GPUCode** [comment]
Hm actually I can reproduce the invalid register writes now, strange. Will look into it

**Comment by GPUCode** [comment]
These invalid reads are pretty strange, the only occur in 2 command lists and are from the first 2 commands with the majority being from extra parameters. The name 0xFEAD also makes me suspect that the game is intentionally using this value for something? The rest of the command list is fine and even calls other nested command lists

**Comment by GPUCode** [comment]
All there games use nested command lists quite heavily and they seem to jump around from cmd buffer 0 to cmd buffer 1 and back. Mario Golf specifically has a ton of these nested command lists. The invalid commands come from one specific one at address 0x180c8000. However whats interesting is that while this command list is mostly garbage, the end of it still jumps around to cmd list 1 and execution continues normally. I wonder if having a few garbage entries is a bug from us or the games fault

**Comment by GPUCode** [comment]
Also @Dragios I tested Mini Mario and Friends on Nightly 2058 with Critical log filter and the game still softlocks, am I missing some other setting?

**Comment by Dragios** [comment]
> Also @Dragios I tested Mini Mario and Friends on Nightly 2058 with Critical log filter and the game still softlocks, am I missing some other setting?

@GPUCode I'm not sure what is going on, but if your game did in fact soft lock, changing the log filter will not affect it. For my case is the game slow down to like 2-3FPS hence setting log filter makes a difference here.

Video as proof:
https://github.com/citra-emu/citra/assets/18496868/2b248881-ce8f-4511-93cc-0d8b36aacfdf

File: src/video_core/renderer_vulkan/pica_to_vk.h (hunk 7 above)
Comment (FearlessTobi) [reply src/video_core/renderer_vulkan/pica_to_vk.h:60]: I'm still able to access our telemetry data fine.
I can send you the credentials if you need them.

**Comment by GPUCode** [comment]
I see, might be a difference between operating systems

**Comment by GPUCode** [comment]
Im really starting to think that the game is sending garbage lists intentionally for whatever reason. The reason all these games were fixed isnt actually because of the command list accuracy improvements but due to the bit-width change in extra data parameter length. On master that the field is 11-bit, this would trigger a lot of invalid extra data parameters and incorrectly treat the jump command as a garbage write, skipping a bunch of command lists at best or slowing down to a halt at worst. Now it stops exactly at a point where the jump is executed properly and the extra geometry is rendered correctly. I find it hard to see how this can be achieved accidentally, though never say never I guess

**Comment by FearlessTobi** [comment]
This PR is likely causing https://github.com/citra-emu/citra/issues/7287.

**Comment by Steveice10** [comment]
clang-format is failing

File: src/citra_qt/debugger/graphics/graphics.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Steveice10) [reply src/citra_qt/debugger/graphics/graphics.cpp:79]: Could maybe add hooks in the main Qt UI logic to (un)register when the game is stopped or started if open.

[the remark this answers is not in the document]
Comment (GPUCode) [reply src/citra_qt/debugger/graphics/graphics.cpp]: I have pushed a possible solution to a new commit

**Review by Steveice10 — APPROVED** [review]

**Comment by Dragios** [comment]
@GPUCode I found a regression on Android build with this PR. Bottom screen seems to glitch out for some Capcom games.
| Screenshot_20240110_210348_Citra  | Screenshot_20240110_210423_Citra   |
|---|---|

For _Phoenix Wright: Ace Attorney - Dual Destinies_ game, it can fix itself if I enter Extras and DLC menu (both top and bottom screen will change content) and exit back to main menu

Unfortunately for _The Great Ace Attorney_ (English patched) game it continue to glitch out even if I try exit back to main menu. It will exhibit different glitches as shown in screenshot below:

| Screenshot_20240110_210553_Citra | Screenshot_20240110_210720_Citra |
|---|---|

**Comment by GPUCode** [comment]
Is the issue fixed by the linked PR? Does it affect only android and not desktop? Are you using OpenGL or Vulkan?
